### ISSUE TITLE
[C++]How to correctly accelerate Parquet read using the FileReader::RowGroup(i)->ReadTable interface

### ISSUE THREAD

**Issue by renyd123** [apache/arrow#42097]
### Describe the usage question you have. Please include as many useful details as  possible.

I encountered a performance issue when calling the FileReader::RowGroup(i)->ReadTable interface in a multi-threaded environment. For a RowGroup with 1 million rows, when the number of threads is 1, FileReader::RowGroup(i)->ReadTable takes around 40ms. However, as the number of threads increases, the execution time also increases gradually. When the number of threads reaches 16, the execution time reaches around 250ms.
This is reflected in the overall file parsing performance, where using 16 threads is only about 5 times faster than using a single thread.
Is this issue caused by locking or I/O in the interface? Are there any other methods for multi-threaded parsing?

### Component(s)

C++

**Comment by mapleFU** [comment]
> For a RowGroup with 1 million rows, when the number of threads is 1, FileReader::RowGroup(i)->ReadTable takes around 40ms. However, as the number of threads increases, the execution time also increases gradually. When the number of threads reaches 16, the execution time reaches around 250ms.

Are you read file on object store or Local FileSystem? I guess the task is a CPU-bound task, and may I ask the CPU count you're using?

**Comment by mapleFU** [comment]
I've check the code here: https://github.com/apache/arrow/blob/03a960de86d0c3ce8bf299724c94be5b291d85b4/cpp/src/parquet/arrow/reader.cc#L1247-L1284

CPU Executor is used when `use_thread` enabled ( C++ disable use_thread by default ), and per-thread calls a `ReadColumn` to emit the column out, generally it couldn't get slower here. I guess maybe that's because contention on CPU thread or other reason 🤔, maybe I can draft some implemention later this week

**Comment by mapleFU** [comment]
Besides, the concurrency is in column-level, which means in per rowgroup, it decode multiple leaf-column concurrently, how many columns do your file has?

**Comment by renyd123** [comment]
> I've check the code here:
> 
> https://github.com/apache/arrow/blob/03a960de86d0c3ce8bf299724c94be5b291d85b4/cpp/src/parquet/arrow/reader.cc#L1247-L1284
> 
> CPU Executor is used when enabled ( C++ disable use_thread by default ), and per-thread calls a to emit the column out, generally it couldn't get slower here. I guess maybe that's because contention on CPU thread or other reason 🤔, maybe I can draft some benchmark later this week`use_thread``ReadColumn`

I deliberately turned off 'use_thread' and instead manually called 'reader->RowGroup(i)->ReadTable(&table)' with multiple threads just like:
`
        threads.push_back(std::thread([&reader, i]() {
            reader->RowGroup(i)->ReadTable(&table);
        }));`

**Comment by mapleFU** [comment]
Aha. so a thread only handle one `ReadTable` call? Basically, if encryption is not enabled, and file doesn't on object-store, `ReadTable` is probably a cpu-bound task...I'll try to repro on my 3800x...

**Comment by renyd123** [comment]
> Aha. so a thread only handle one `ReadTable` call? Basically, if encryption is not enabled, and file doesn't on object-store, `ReadTable` is probably a cpu-bound task...I'll try to repro on my 3800x...

yes,  16 threads handle 16 RowGroups in parallel. And columns are processed serially in each thread.

**Comment by thisisnic** [comment]
Closing as stale usage question with no recent activity.
